### PR TITLE
CI Testing: Set Memory Layout On Purpose 

### LINKED ISSUE #26502: Copy and grouby-sum changes column-major-ordering to row-major-ordering with huge impact on performance
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd

print("pandas version: ", pd.__version__)

array = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
array.flags
print("Numpy array is C-contiguous: ", data.flags.c_contiguous)

dataframe = pd.DataFrame(array, index = pd.MultiIndex.from_tuples([('A', 'U'), ('A', 'V'), ('B', 'W')], names=['dim_one', 'dim_two']))
print("DataFrame is C-contiguous: ", dataframe.values.flags.c_contiguous)

dataframe_copy = dataframe.copy()
print("Copy of DataFrame is C-contiguous: ", dataframe_copy.values.flags.c_contiguous)

aggregated_dataframe = dataframe.groupby('dim_one').sum()
print("Aggregated DataFrame is C-contiguous: ", aggregated_dataframe.values.flags.c_contiguous)

## Output in Jupyter Notebook
# pandas version:  0.23.4
# Numpy array is C-contiguous:  True
# DataFrame is C-contiguous:  True
# Copy of DataFrame is C-contiguous:  False
# Aggregated DataFrame is C-contiguous:  False

```
#### Problem description

pandas.DataFrame.copy return a copy with underlying numpy array with the major-ordering changed from C to F if initially was C order

pandas.DataFrame.groupby.sum return a a DataFrame with underlying numpy array with the major-ordering changed from C to F if initially was C order on the DataFrames values.

This impacts performance on aggregation substantially! I observer a decrease from 17 seconds to 5 min 46 seconds when aggregating on the index on a 45023 times 100000 scenarios dataset.

Copy and Aggregation operations need to preserve the major-order on the underlying data! 

#### Expected Output

```python
# Output in Jupyter Notebook
 pandas version:  0.23.4
 Numpy array is C-contiguous:  True
 DataFrame is C-contiguous:  True
 Copy of DataFrame is C-contiguous:  True
 Aggregated DataFrame is C-contiguous:  True
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-862.14.4.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 3.2.3
pip: 19.0.2
setuptools: 40.8.0
Cython: 0.22
numpy: 1.15.4
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.6.3
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: 2.2.2
openpyxl: 2.4.7
xlrd: 1.1.0
xlwt: None
xlsxwriter: 1.0.2
lxml: None
bs4: None
html5lib: None
sqlalchemy: 1.3.3
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by TannhauserGate42**
- [ ] closes #26502
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 import warnings
 
 import numpy as np
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py]: This import would need to go below with other core imports.

**Comment by jbrockmendel** [comment]
Is the idea here to let the user specify a preferred default layout?  That could go in e.g. pd.options.

Needs tests.

**Comment by TannhauserGate42** [comment]
> Is the idea here to let the user specify a preferred default layout? That could go in e.g. pd.options.
> 
> Needs tests.

I will provide some. First, I need to succeed to check consistency and get through the unittests ... it might have wide reaching dependencies. Need to get the unittest suite running properly on my machines. ... not a pro, sorry. 

I need to understand some tests first. changing memory layout might have consequences especially in conjunction with pytables tests, should there be implicit assumptions on the memory layout given so far. I hope, it does not. ... but hope is not a method.

**Comment by TomAugspurger** [comment]
Just a short FYI, this may not be something we can actually guarantee going
forward. There's a possibility that pandas' internal representation will
change from a mixture of 1D and 2D arrays
to just 1D arrays (though the 1D arrays may be views on a 2D array). All
that is quite uncertain though.

**Comment by jorisvandenbossche** [comment]
@TannhauserGate42 can you indicate how this is different to the previous PR (https://github.com/pandas-dev/pandas/pull/26529)? If there is no difference: practical point of feedback: it is not needed to open new PRs when updating based on feedback.

But, I agree with @TomAugspurger that we might not necessarily want to include something like this. So I would recommend to first focus on actually discussing this before putting too much effort in a PR (although for the discussion, having benchmarks might be needed, which might need an initial implementation).

**Comment by TannhauserGate42** [comment]
> @TannhauserGate42 can you indicate how this is different to the previous PR (#26529)? If there is no difference: practical point of feedback: it is not needed to open new PRs when updating based on feedback.
> 
> But, I agree with @TomAugspurger that we might not necessarily want to include something like this. So I would recommend to first focus on actually discussing this before putting too much effort in a PR (although for the discussion, having benchmarks might be needed, which might need an initial implementation).

Thanks. 

I have the testsuite now running locally and will try another concept for our own purpose. Thanks for letting me use your CI for testing. 

I aggree, that my implementation is not the one to go forward and is not meant as such. At the same time it shows me, that pandas implicitly relies on some numpy memory layout and lots of stuff would brake, if these implicit assumptions (like np-defaulting to C-order) would not hold anymore.

It would be good, if pandas would allow to set the memory layout or at least would be consistently always use the same (which I showed is not true).

Unfortunatly, we have severer problems with performance now and with a quick fix I went down from 1:25 hours to 17 minutes. All because aggregation on the one and same dataframe takes about 5 minutes with F-order and about 20 with C-order memory layout. 

I leave you with the question: Why does pandas indexing depend on the memory layout of numpy ndarrays?

**Pull request closed without merging**